### ISSUE TITLE
Return code of 204

### ISSUE THREAD

**Issue by Emily-Jiang** [eclipse/microprofile-health#3]
In the spec, the return code 204 means service UP. However, in cloud foundry, it does not understand the return code 204. Shall we use just use 200 to mean UP? It is fine to have different payload with the same return code. Thoughts?

**Comment by heiko-braun** [comment]
In the spec 204 means that no health check procedures are installed. In this situation the outcome should still signal a positive result, but no content. 

Do you have any details on the cloud foundry health check mechanism and it's supported protocol elements? I think supporting it would be good, but for that we need more details.

**Comment by Emily-Jiang** [comment]
https://docs.cloudfoundry.org/devguide/deploy-apps/healthchecks.html#understand-healthchecks has the details.

**Comment by starksm64** [comment]
The referenced document describes the http health check type as:

> The http health check performs a GET request to the configured http endpoint. When the health check receives an HTTP 200 response, the app is declared healthy. We recommend using the http health check type whenever possible. A healthy HTTP response ensures that the web app is ready to serve HTTP requests. The configured endpoint must respond within 1 second to be considered healthy.

So it is just a simple probe based on an HTTP_OK response. The other network oriented health check is a port check, and this simply verifies that a connected state can be reached. There are no payloads in their health check protocol.

**Comment by heiko-braun** [comment]
Previously we've used 204 (no content) to make it clear that no health check procedures are installed into the runtime.  But I am fine with changing the default status code to `200` to support cloud foundry and just not return any payload. 

Skimming through the HTTP spec I think this is acceptable:

```
 Aside from responses to CONNECT, a 200 response always has a payload,
   though an origin server MAY generate a payload body of zero length.
```

**Comment by heiko-braun** [comment]
Keep in mind, when we change this to `200`, the TCK tests lack the ability to verify if procedures are installed. Currently the distinction between `200` and `204` is used to verify if the procedures have been correctly installed into the runtime. We may need an equivalent to this and I think it would require a response payload structure (to be defined) that indicates the lack of procedures.

**Comment by Emily-Jiang** [comment]
Just verify whether the payload is empty or not. If yes, it means 204 in the original structure. This matches to CF exception. Right?

**Comment by starksm64** [comment]
A Content-Type is only a SHOULD header in the HTTP spec, so I think we can look to mimic the 204 behavior to looking for the lack of a Content-Type header + 200 == 204. The question will be can we prevent a Content-Type header from being added on the server side.

Otherwise, as @Emily-Jiang says, the lack of a payload would just have to be viewed as indicating that there are no configured procedures for this initial release. As @heiko-braun indicates, we may need to revise the payload for this situation going forward.

**Comment by heiko-braun** [comment]
Both approaches work for me.

**Comment by starksm64** [comment]
There is agreement on this approach. @Emily-Jiang, @heiko-braun and myself talked about using a 200 response status and not include a Content-Type header if there are no configured procedures, while if there are installed procedures, the Content-Type would be application/json.

**Comment by heiko-braun** [comment]
I think a more idiomatic way (i.e. in line with the remaining protocol) would be to return 200 and a body of:

```
{
  "outcome": "UP",
  "checks": []
}
```

We only used to return 204 to indicate that no procedures are installed, but the system should be treated as `UP`. I think the empty checks array, plus overall outcome `UP` reflects this perfectly.

I'd rather prefer this approach to fiddling with contents types, etc

**Comment by heiko-braun** [comment]
The companion document [1] (see Section JSON Payload and status codes) needs to be updated as well to reflect the changes discussed here.

[1] : https://github.com/eclipse/microprofile-health/blob/master/spec/src/main/asciidoc/protocol-wireformat.adoc#with-procedures-installed-into-the-runtime

**Comment by starksm64** [comment]
Resolved with #67
